### ISSUE TITLE
All sys.path modules are loaded at the beginning of execution 

### ISSUE THREAD

**Issue by rsanchezgarc** [I2PC/scipion#1830]
When you start scipion, it reaches 

```
    @classmethod
    def __getSubclasses(cls, submoduleName, BaseClass):
```
in /home/rsanchez/app/scipion/pyworkflow/plugin.py

Thus, heavy modules (as tensorflow) are loaded in splite of never being used.

**Comment by delarosatrevin** [comment]
Hi @rsanchezgarc, thanks for reporting this. I have also observed a long time opening Scipion, I guess for the same reason of a lot of libraries loading (unnecessary). I think this will be solved in the pluginized version, where we are not loading all protocols by default (and then all required Python packages), still, we should keep an eye on this, to avoid the same mistake.

**Comment by pconesa** [comment]
If I'm right, @rsanchezgarc is referring to the pluginized version, and this is in the process of discovering plugins, where we are loading every python module present in the installation to check if it is a plugin.

**Comment by delarosatrevin** [comment]
Opps...yes, this could slow down everything. @pconesa What do you think about discovering plugins once and then use a cache value? Instead of discovering every time Scipion runs? The challenge here is to make sure when a new plugin is installed, run the discovery process....maybe we need to brain-storm a bit about this. This is an important issue.

**Comment by pconesa** [comment]
There is this....https://packaging.python.org/guides/creating-and-discovering-plugins/

**Comment by delarosatrevin** [comment]
@yaizar I don't like the idea of enforcing the python module to be a plugin or pip package...this can be troublesome while developing the module and all the plugin structure is not done.

**Comment by delarosatrevin** [comment]
Maybe iterating through all modules path and check for files is not good neither...I'm not sure of the standard python modules if there are in eggs or other pre-compiles forms...how the path are handled. 

What about having a cached file with "known" plugins? We can keep the current way as to "discover" everything and maybe to update the cache. Having the "cache plugins file" could also be interesting for development, were one could play with it and activate/deactivate plugins for testing purposes.

**Comment by yaizar** [comment]
that sounds VERY similar to the plugins.json file. What about when you want to add your plugin from the python path? or when you make a link to test your plugin in development? does the developer update the cache manually?

**Comment by pconesa** [comment]
Any cache mechanism, should be carefully reset....not only in development mode, but when installing new plugins, at least in this case.

**Comment by delarosatrevin** [comment]
Yes, that is true....

By the way...have someone checked when the function Domain.getPlugins is invoked?
I think it is not always...if it is mainly from the GUI, maybe could be a way to push this to a worker thread?

**Comment by pconesa** [comment]
And if at some point we decide to have .eggs, and this does not work....then we can go for something more elaborated like the cached you are suggesting.

**Comment by pconesa** [comment]
@delarosatrevin, haven't check that...but I guess is happening always, not only GUI ....when you run a protocol, it should happen too. Right?

**Comment by delarosatrevin** [comment]
Hum...yes, It also should be loaded when running a protocol...but I guess that the computing will be much longer than loading time...@rsanchezgarc, has you experienced this delay while computing? or in the GUI?

Anyway...we really don't need all protocols plugins...when running a given protocols...but changing this seems much more complicated right now...but could be another option.

**Comment by rsanchezgarc** [comment]
I hace experienced it in both. Although it is not as important as when
opening Scipion, it si noticiable. Moreover. Some warnings are shown. I
think that a message like "loading tensor devices" when running an subset
protocolo could be avery misleading.

Ruben

**Comment by delarosatrevin** [comment]
Yep, it is. Thanks for the feedback @rsanchezgarc!

**Comment by delarosatrevin** [comment]
Well...thinking again about loading just the needed modules (without iterating over all of them) for discovering plugins...will not help if the developer in a protocol put the tensorflow as an import...I think we discussed a little bit about it during the developers meeting and a way could be to make a separate script for doing the real computation...and then treat this as an external program from Scipion perspective. (it could also be good in that way to isolate python library versions and other dependencies).

**Comment by pconesa** [comment]
Yep, that is true. What about local imports.... at least if tensorflow is a local import, we are saving that, rigth?

**Comment by delarosatrevin** [comment]
Yes, either local imports or separate scripts...I think this will be the easiest solution right now...and we don't need to modify nothing about the current behavior. In general, I would prefer the script way...because we could also define different dependencies and libraries (of the script) with environments, without adding dependencies to the plugin itself.

**Comment by dmaluenda** [comment]
I don't think so! We realized (with @rsanchezgarc ) that is not necessary to import the heavy module (like tensorflow, keras) anywhere inside Scipion to get this issue, just having the module installed is enough to get this issue because, it is also loaded since we inspect all the available modules in site-package in order to see if it match with a Scipion plugins.

We double check that even without import "keras" anywere in the code, the "Using tensorflow backend" mesage still appearing at the very early steps of launching Scipion.

**Comment by delarosatrevin** [comment]
@dmaluenda I think we should not install heavy modules (tensorflow, keras, etc) in the Scipion python...but maybe define a virtualenv or similar...and install necessary dependencies there. In this way, we clearly separate the Scipion's python from the Plugin's python...that is already happening for other plugins such as crYOLO and Topaz...so, I will suggest to consider the possibility to define an environment for Xmipp and avoid installing its dependencies into the Scipion python.

**Comment by yaizar** [comment]
Regarding the computation scripts a bit of the same, that's fine but it's a different issue. I think we're a bit late to start implementing a cache, so I'm opposed to this solution. I think assuming that a scipion plugin has a protocols folder or a protocols.py or something along these lines is a more feasible solution to have up and running before we freeze for the release. What can a plugin do if you don't have a protocols folder/module?

**Comment by pconesa** [comment]
@delarosatrevin, I agree with your solution.....but it seems to me a lot of work, and due to the time constraints, maybe there is not enough time for that...or...maybe you see it much easier that I do and is even a "quick refactor"?

**Comment by delarosatrevin** [comment]
@yaizar, the script thing is a different issue, but it is related to this one...already we are checking for a given structure in the plugins (Plugin class), but we need to go and import all ones in the Python. Looking for protocols.py is not a good idea...in Python, a module could be a single file or a subfolder...moreover...I/O operations (e.g file exists, etc) can be slow in some setups where the installation is via network or so. 

@pconesa I don't know how much change is this...maybe just in Xmipp to ask for a Python (or environment) and install there these dependencies (also only for some programs/protocols...not for the core). I see this more like a scipion-em-xmipp problem. Of course, I'm not against improving our performance in general...but as @yaizar said, cache or other elaborated solutions seems a bit in a hurry now for the release. 

So, my proposal is:
1) Do nothing...if you have Scipion and install Xmipp plugin dependencies, you know it will hit us and will be slow
2) Maybe (but this is up to the Xmipp developers) considering moving the dependencies to a script and using virtualenv.
2.1) Maybe still is worth moving tensorflow/keras to script/local imports...and then if I install Scipion and Xmipp, but not these heavy modules...I can use Scipion/Xmipp but without the programs that use these dependencies.
3) For Scipion core in the future, try to improve how plugins are discovered...

Still, as a general rule, I think we should separate Scipion dependencies from Plugins (including Xmipp) dependencies.

**Comment by delarosatrevin** [comment]
How difficult would be to test:
1) Create a virtualenv within the Scipion python and install Xmipp deps there
2) Activate this vritualenv when running such protocols that require to import/use heavy modules
3) Do don't install these heavy modules in the root Scipion python

???

**Comment by dmaluenda** [comment]
I strongly agree with the 3rd point, and I compromise that if any heavy module is imported in the root Scipion python code I will move it to a local scope.

I think and I hope that Tensorflow and Keras (among others heavy modules) are not hard dependency for Xmipp, but they are dependencies only if you want to use a protocol that it needs it.

**Comment by dmaluenda** [comment]
I don't exactly know how to define and work virtualenv, therefore I can't get in charge of it for this new version, but I am very motivated to learn about it to implement it after this next release.

**Comment by yaizar** [comment]
@delarosatrevin OK maybe we have to talk more somewhere else about that script because I guess I'm not understanding what you're proposing. What I'm understanding is that you and  Pablo are talking about how to load heavy libraries at execution time (i.e. in the protocols) and that this coiuld be done in a computation script. Am I right? We still shouldnt try to import everything.
Why do you think it is gonna be more costly to check if a module exists within module X than to actually try to import module X in an installation via network? I honestly doubt it...
The fact that it can be a file or a subfolder is irrelevant because like you say, it behaves like a module. So we would check for the existence of a module using some library like pkg_util. This isnt incompatible with xmipp using a different python in the future, but I still think that's a separate topic.

**Comment by delarosatrevin** [comment]
@yaizar I totally agree that maybe should discuss this further in other occasion. Here all softwares are loaded via modules, from a central server through the network...and Scipion already take a while to load...even without the issues of the heavy modules. I don't totally get your point of changing the current way of loading the modules by a similar one...but let's discuss this in another moment.

**Comment by pconesa** [comment]
Thanks @delarosatrevin , 1 will happen to the next release, unless we/xmipp do 2. That is my concern.

@rsanchezgarc , @dmaluenda how do you see @delarosatrevin sugggestion?

**Comment by rsanchezgarc** [comment]
I think that virtual environments is the most robust way to solve these
problems.

**Comment by yaizar** [comment]
yeah nevermind, I thought pkgutil had a tool to walk submodules without executing __init__ but no, it would have to be done checking the path. However one line of code and the assumption that a plugin has a protocols folder or file (which all do atm) would solve this for now.

**Comment by pconesa** [comment]
Fixed in scipion 3.0 using entry points
